### PR TITLE
CA-1109 Adding logging

### PR DESCRIPTION
**Pull request by gpolumbo-broad**
I _think_ the issue has to do with the fact that "acquiring a lock" on a FenceServiceAccount record in Datastore can fail because the record is either already locked, or if the Datastore transaction throws any kind of exception for any reason, which, according to Google's documentation, is a possible thing:  https://cloud.google.com/appengine/docs/standard/python/datastore/transactions#using_transactions

At this time, we don't know if this is what's happening (because any exceptions here are swallowed) or if something else is going on.  Bond could really use a bunch more logging in general, so I added several log statements in the whole "get Key" flow that hopefully will help to debug this (and future) issue(s).  

Also started adding support for more configurable logging.  There's surely even more we can do to make the logging easier to use to ensure that you're using the right logger.   I think we should also adjust the formatter to log in JSON format to aid Google Cloud Logging in parsing our log messages.  But this is just a first step so we can try to resolve the issue at hand.  

Also needed to update the `google-cloud-logging` library which then triggered a cascade of other libraries that needed to be updated.

Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've followed [the instructions](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/DataBiosphere/bond/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary. Documentation PRs only need 1 thumb.
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: bond_app/fence_api.py [hunk 1]
```diff
@@ -18,10 +18,16 @@ def get_credentials_google(self, access_token):
         """
         headers = {'Authorization': 'Bearer ' + access_token}
         result = requests.post(url=self.credentials_google_url, headers=headers)
+        # Logging the `result.content` here as part of debugging/troubleshooting for:
+        # https://broadworkbench.atlassian.net/browse/CA-1109
+        # Logging of the `response.content` can probably be removed after we have resolved this issue
+        logging.info("Getting new Service Account JSON Key from Fence via - request: POST {} - status code: {} - body: {}" \
+                     .format(self.credentials_google_url, result.status_code, result.content))
```
Comment (dvoet) [inline bond_app/fence_api.py]: The content being logged here is a service account key with access to protected data! Definitely don't log that.

File: bond_app/fence_token_storage.py [hunk 2]
```diff
@@ -105,6 +106,15 @@ def _fetch_and_cache_service_account(self, provider_user, prep_key_fn, fence_fet
...
         """
+        This thing can return FALSE for 2 different reasons!  And those "FALSE" responses might mean two very different
+        things.  We are currently investigating: https://broadworkbench.atlassian.net/browse/CA-1109
+
+        Either we were unable to successfully lock the FenceServiceAccount record because it is already locked, OR
+        something else happened.  In the latter case, the transaction was successful, but the call to grab the lock
+        still threw an Exception for some reason.  When this happens, do we need to do something different to handle
+        that scenario?  Do we do something different based on the type of the exception that was thrown?  Do we do
+        nothing and just let the exception bubble up?  I think it all depends on the type of Exception that gets thrown,
+        but at the moment, we don't know what those exceptions might be nor what they mean.
```
Comment (dvoet) [inline bond_app/fence_token_storage.py]: It looks like the comments below cover this case. In any case, I think it is inappropriate to commit debug comments like this.

File: bond_app/fence_token_storage.py [hunk 3]
```diff
@@ -161,8 +172,10 @@ def _lock_fence_service_account(self, fsa_key):
         if fence_service_account is None:
             fence_service_account = FenceServiceAccount(key=fsa_key, update_lock_timeout=update_lock_timeout)
         elif fence_service_account.update_lock_timeout and fence_service_account.update_lock_timeout > datetime.datetime.now():
+            logging.info("Could not obtain lock on {} because it is already locked".format(fence_service_account))
```
Comment (dvoet) [inline bond_app/fence_token_storage.py]: I think this is debug level

File: bond_app/fence_token_storage.py [hunk 4]
```diff
@@ -161,8 +172,10 @@ def _lock_fence_service_account(self, fsa_key):
...
             return False
         else:
             fence_service_account.update_lock_timeout = update_lock_timeout
         fence_service_account.put()
+        logging.info("Successfully locked FenceServiceAccount Record: {}".format(fence_service_account))
```
Comment (dvoet) [inline bond_app/fence_token_storage.py]: I think this is debug level

**Review by dvoet — CHANGES_REQUESTED** [review]

**Review by andy7i — APPROVED** [review]
These look like potentially very helpful changes, bubbling up the exception and enabling local logging. 👍 with the changes doug requested.

File: bond_app/fence_api.py [hunk 5]
```diff
@@ -18,10 +18,16 @@ def get_credentials_google(self, access_token):
         """
         headers = {'Authorization': 'Bearer ' + access_token}
         result = requests.post(url=self.credentials_google_url, headers=headers)
+        # Logging the `result.content` here as part of debugging/troubleshooting for:
+        # https://broadworkbench.atlassian.net/browse/CA-1109
+        # Logging of the `response.content` can probably be removed after we have resolved this issue
```
Comment (dvoet) [inline bond_app/fence_api.py]: already gone

**Review by dvoet — APPROVED** [review]

**Review by dvoet — APPROVED** [review]

File: bond_app/fence_api.py [hunk 6]
```diff
@@ -18,10 +21,34 @@ def get_credentials_google(self, access_token):
...
+        :param json_string:
+        :param name:
+        :param redact_keys:
+        :return:
+        """
+        try:
+            scrubbed_json_result = json.loads(json_string)
+            for redact_key in redact_keys:
+                scrubbed_json_result[redact_key] = "[REDACTED]"
+            logger.debug("Scrubbed {}: {}".format(name, scrubbed_json_result))
```
Comment (andy7i) [inline bond_app/fence_api.py:46]: verified that it works locally!

```
2021-01-15 10:46:38,738 - bond_app.fence_api - DEBUG - Getting new Service Account JSON Key from Fence via - request: POST https://us-central1-broad-dsde-dev.cloudfunctions.net/user/credentials/google - status code: 200 -
2021-01-15 10:46:38,739 - bond_app.fence_api - DEBUG - Scrubbed Service Account Key JSON: {'auth_provider_x509_cert_url': 'https://www.googleapis.com/oauth2/v1/certs', 'auth_uri': 'https://accounts.google.com/o/oauth2/auth', 'client_email': 'mock-provider-user-service-acc@broad-dsde-dev.iam.gserviceaccount.com', 'client_id': '102327840668511205237', 'client_x509_cert_url': 'https://www.googleapis.com/robot/v1/metadata/x509/mock-provider-user-service-acc%40broad-dsde-dev.iam.gserviceaccount.com', 'private_key': '[REDACTED]', 'private_key_id': '9d41d414182a8ff73f944ee89d67c605d1821589', 'project_id': 'broad-dsde-dev', 'token_uri': 'https://oauth2.googleapis.com/token', 'type': 'service_account'}
2021-01-15 10:46:38,746 - werkzeug - INFO - 127.0.0.1 - - [15/Jan/2021 10:46:38] "GET /api/link/v1/dcf-fence/serviceaccount/key HTTP/1.1" 200 -
```

**Review by andy7i — APPROVED** [review]
👍  the comments help a lot to describe what's going on
did some light testing and saw the logs as expected

**Comment by gpolumbo-broad** [comment]
jenkins retest

**Pull request merged**